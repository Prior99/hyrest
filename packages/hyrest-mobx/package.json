{
  "name": "hyrest-mobx",
  "version": "0.0.0",
  "description": "A hybrid REST library for the frontend and backend. Mobx connector package.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise",
    "form",
    "field",
    "mobx"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "typedoc": "^0.11.1",
    "typescript": "^3.0.1"
  },
  "dependencies": {
    "bind-decorator": "^1.0.11",
    "hyrest": "^0.8.0",
    "mobx": "^5.0.3",
    "reflect-metadata": "^0.1.12"
  }
}
