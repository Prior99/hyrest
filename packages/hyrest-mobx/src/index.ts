export * from "./field";
export * from "./decorators";
export * from "./types";
export * from "./validation-status";
