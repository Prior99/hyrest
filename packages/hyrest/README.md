# Hyrest

[![npm](https://img.shields.io/npm/v/hyrest.svg)](https://www.npmjs.com/package/hyrest)
[![Build Status](https://travis-ci.org/Prior99/hyrest.svg?branch=master)](https://travis-ci.org/Prior99/hyrest)
[![Coverage Status](https://coveralls.io/repos/github/Prior99/hyrest/badge.svg?branch=master)](https://coveralls.io/github/Prior99/hyrest?branch=master)

Hyrest is a hybrid REST framework for both the client and the server.

The idea is to define routes using decorators and use them to both serve the REST endpoint
and call them from the frontend. When developing both server and client in the same repository
or sharing a common library with all endpoints between the both, a call to a REST endoint
is transparent, type-safe and as easy calling a method.

## Usage

Please refer to the [resources](#resources) for getting you started.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest/)
