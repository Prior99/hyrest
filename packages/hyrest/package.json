{
  "name": "hyrest",
  "version": "0.10.0",
  "description": "A hybrid REST library for the frontend and backend.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "@types/invariant": "^2.2.29",
    "typedoc": "^0.12.0",
    "typescript": "^3.0.3"
  },
  "dependencies": {
    "http-status-codes": "^1.3.0",
    "invariant": "^2.2.2",
    "isomorphic-fetch": "^2.2.1",
    "path-to-regexp": "^2.4.0",
    "reflect-metadata": "^0.1.10"
  }
}
