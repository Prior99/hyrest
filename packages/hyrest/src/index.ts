import "path-to-regexp";

export * from "./all-keys";
export * from "./answers";
export * from "./authorization";
export * from "./configure";
export * from "./controller";
export * from "./converters";
export * from "./is-browser";
export * from "./last-call";
export * from "./parameters";
export * from "./processed";
export * from "./route";
export * from "./schema-generator";
export * from "./scope";
export * from "./transform";
export * from "./types";
export * from "./validation";
export * from "./validators";

import * as dataType from "./converters";
export const DataType = dataType;

import * as http from "./answers";
export const Http = http;
