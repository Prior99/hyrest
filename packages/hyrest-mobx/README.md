# Hyrest MobX

<img align="right" width="200" height="200" src="https://github.com/Prior99/hyrest/raw/master/logo/hyrest-logo-400px.png">

[![npm](https://img.shields.io/npm/v/hyrest-mobx.svg)](https://www.npmjs.com/package/hyrest-mobx)
[![pipeline status](https://gitlab.com/prior99/hyrest/badges/master/pipeline.svg)](https://github.com/Prior99/hyrest)
[![coverage report](https://gitlab.com/prior99/hyrest/badges/master/coverage.svg)](https://github.com/Prior99/hyrest)

Hyrest is a hybrid REST framework for both the client and the server.

**This is the MobX frontend utility package.**

After creating and exposing a REST Api using [hyrest](../hyrest) and [hyrest-express](../hyrest-express) it might be useful to re-use the existing models and validation from the server side in the client to store data next to forms and perform validation on it.

## Usage

Let's say a model has been defined which is already used by a controller:

```typescript
import { is, email, length } from "hyrest";
class User {
    @is().validate(email) public email: string;
    @is().validate(length(5, 20)) public name: string;
    @is().validate(length({ min: 10 })) public password: string;
}
```

In order to create a new user, one might want to implement a sign up form. This is where this package can come in handy.

Instead of creating a `@observable public email: string` and so on for each field on the `User` model, simple utilize hyrest-mobx:

```typescript
import * as React from "react";
import { observer } from "mobx-react";
import { field, hasFields, Field } from "hyrest-mobx";

@observer @hasFields()
class Signup extends React.Component {
    @field(User) private user: Field<User>;

    @action.bound private handleSubmit() {
        console.log(user.value); // Will log the unwrapped user.
    }

    public render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <input type="text" {...this.user.nested.email.reactInput} />
                <input type="text" {...this.user.nested.name.reactInput} />
                <input type="password" {...this.user.nested.password.reactInput} />
                {
                    this.users.errors.map((err, index) => <p key={index}>{err}</p>)
                }
                <button disabled={!this.user.valid}>Signup</button>
            </form>
        );
    }
}
```

The decorator `@field` will mark the property as a field with the type `User`.

After marking the class with `@hasFields`, new `Field`s will automatically be injected upon creation.

The injected values already utilize Mobx under the hood and hence no `@observable`s are no longer necessary.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest-mobx/)
