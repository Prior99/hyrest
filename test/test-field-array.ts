import { is, specify } from "hyrest";
import { field, hasFields, Field, ValidationStatus } from "hyrest-mobx";

class A {
    @is()
    public something: string;
}

class List {
    @is() @specify(() => A)
    public items: A[];
}

@hasFields(() => ({}))
class Container {
    @field(List) public list: Field<List>;
    @field(List) public anotherList: Field<List>;
}

let instance: Container;

beforeEach(() => instance = new Container());

test("with the status being valid", async () => {
    await instance.list.update({ items: [{ something: "a string" }] });
    expect(instance.list.nested.items.status).toBe(ValidationStatus.VALID);
    expect(instance.list.nested.items.valid).toBe(true);
    expect(instance.list.nested.items.invalid).toBe(false);
    expect(instance.list.nested.items.unknown).toBe(false);
    expect(instance.list.nested.items.inProgress).toBe(false);
});

test("with the status being invalid", async () => {
    await instance.list.update({ items: [{ something: 9 } as any] });
    expect(instance.list.nested.items.status).toBe(ValidationStatus.INVALID);
    expect(instance.list.nested.items.valid).toBe(false);
    expect(instance.list.nested.items.invalid).toBe(true);
    expect(instance.list.nested.items.unknown).toBe(false);
    expect(instance.list.nested.items.inProgress).toBe(false);
});

test("with the status being unkown", async () => {
    expect(instance.list.nested.items.status).toBe(ValidationStatus.UNKNOWN);
    expect(instance.list.nested.items.valid).toBe(false);
    expect(instance.list.nested.items.invalid).toBe(false);
    expect(instance.list.nested.items.unknown).toBe(true);
    expect(instance.list.nested.items.inProgress).toBe(false);
});

test("with the status being unkown", async () => {
    await instance.list.nested.items.add({ something: "a string" });
    instance.list.nested.items.at(0).nested.something.update("a string");
    expect(instance.list.nested.items.status).toBe(ValidationStatus.IN_PROGRESS);
    expect(instance.list.nested.items.valid).toBe(false);
    expect(instance.list.nested.items.invalid).toBe(false);
    expect(instance.list.nested.items.unknown).toBe(false);
    expect(instance.list.nested.items.inProgress).toBe(true);
});

test("resetting", async () => {
    await instance.list.update({ items: [{ something: "a string" }] });
    expect(instance.list.value).toMatchSnapshot();
    expect(instance.list.status).toBe(ValidationStatus.VALID);
    await instance.list.nested.items.reset();
    expect(instance.list.value).toMatchSnapshot();
    expect(instance.list.status).toBe(ValidationStatus.UNKNOWN);
});

test("length", async () => {
    expect(instance.list.nested.items.length).toBe(0);
    await instance.list.update({ items: [{ something: "a string" }] });
    expect(instance.list.nested.items.length).toBe(1);
});

test("concattenating", async () => {
    await instance.list.update({ items: [{ something: "a string" }] });
    await instance.anotherList.update({ items: [{ something: "another string" }] });
    const array = instance.list.nested.items.concat(instance.anotherList.nested.items);
    expect(array.length).toBe(2);
    expect(array.value).toMatchSnapshot();
});

test("slice", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
            { something: 78 } as any,
        ],
    });
    const subArray = instance.list.nested.items.slice(0, 2);
    expect(subArray.length).toBe(2);
    expect(subArray.value).toMatchSnapshot();
    expect(instance.list.nested.items.status).toBe(ValidationStatus.INVALID);
    expect(subArray.status).toBe(ValidationStatus.VALID);
});

test("indexOf", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.indexOf(instance.list.nested.items.at(2))).toBe(2);
});

test("lastIndexOf", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.lastIndexOf(instance.list.nested.items.at(0))).toBe(0);
});

test("every", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.every(subField => subField.value.something.includes("string"))).toBe(true);
    expect(instance.list.nested.items.every(subField => subField.value.something.includes("1st"))).toBe(false);
});

test("some", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.some(subField => subField.value.something.includes("4th"))).toBe(false);
    expect(instance.list.nested.items.some(subField => subField.value.something.includes("1st"))).toBe(true);
});

test("forEach", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    const strings: string[] = [];
    instance.list.nested.items.forEach(subField => strings.push(subField.value.something));
    expect(strings).toMatchSnapshot();
});

test("map", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.map(subField => subField.value)).toMatchSnapshot();
});

test("filter", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
            { something: "4th string" },
            { something: "5th string" },
        ],
    });
    const newArray = instance.list.nested.items.filter(subField => subField.nested.something.value.includes("th"));
    expect(newArray.length).toBe(2);
    expect(newArray.value).toMatchSnapshot();
});

test("reduce", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    const reduced = instance.list.nested.items.reduce((result, subField) => {
        return `${result}-${subField.nested.something.value}`;
    }, "");
    expect(reduced).toMatchSnapshot();
});

test("reduceRight", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    const reduced = instance.list.nested.items.reduceRight((result, subField) => {
        return `${result}-${subField.nested.something.value}`;
    }, "");
    expect(reduced).toMatchSnapshot();
});

test("find", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    const foundField = instance.list.nested.items.find(subField => subField.nested.something.value === "2nd string");
    expect(instance.list.nested.items.indexOf(foundField)).toBe(1);
    expect(foundField.value).toMatchSnapshot();
});

test("findIndex", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.nested.items.findIndex(subField => subField.nested.something.value === "2nd string")).toBe(1);
});

test("valueAt", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
        ],
    });
    expect(instance.list.nested.items.valueAt(1)).toMatchSnapshot();
});

test("updating with existing fields", async () => {
    await instance.list.update({
        items: [
            { something: "1st string" },
            { something: "2nd string" },
            { something: "3rd string" },
        ],
    });
    expect(instance.list.status).toBe(ValidationStatus.VALID);
    await instance.list.nested.items.update([
        {},
        { something: 78 },
        {},
        { something: "4th string" },
    ] as any[]);
    expect(instance.list.status).toBe(ValidationStatus.INVALID);
    expect(instance.list.value).toMatchSnapshot();
});
