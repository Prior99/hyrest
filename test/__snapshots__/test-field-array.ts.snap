// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concattenating 1`] = `
Array [
  A {
    "something": "a string",
  },
  A {
    "something": "another string",
  },
]
`;

exports[`filter 1`] = `
Array [
  A {
    "something": "4th string",
  },
  A {
    "something": "5th string",
  },
]
`;

exports[`find 1`] = `
A {
  "something": "2nd string",
}
`;

exports[`forEach 1`] = `
Array [
  "1st string",
  "2nd string",
  "3rd string",
]
`;

exports[`map 1`] = `
Array [
  A {
    "something": "1st string",
  },
  A {
    "something": "2nd string",
  },
  A {
    "something": "3rd string",
  },
]
`;

exports[`reduce 1`] = `"-1st string-2nd string-3rd string"`;

exports[`reduceRight 1`] = `"-3rd string-2nd string-1st string"`;

exports[`resetting 1`] = `
List {
  "items": Array [
    A {
      "something": "a string",
    },
  ],
}
`;

exports[`resetting 2`] = `
List {
  "items": Array [],
}
`;

exports[`slice 1`] = `
Array [
  A {
    "something": "1st string",
  },
  A {
    "something": "2nd string",
  },
]
`;

exports[`updating with existing fields 1`] = `
List {
  "items": Array [
    A {
      "something": "1st string",
    },
    A {
      "something": 78,
    },
    A {
      "something": "3rd string",
    },
    A {
      "something": "4th string",
    },
  ],
}
`;

exports[`valueAt 1`] = `
A {
  "something": "2nd string",
}
`;
